walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -40,19 +41,69 @@
     ArrayList<Event> sortedEvents = new ArrayList<>(events);
     Collections.sort(sortedEvents, Event.ORDER_BY_START);
 
+    Collection<Event> optionalEvents = new ArrayList<>();
+
+    ListIterator<Event> eventsIterator = sortedEvents.listIterator();
+    while (eventsIterator.hasNext()) {
+      Event event = eventsIterator.next();
+      
+      // Separates mandatory attendee events and optional attendee events into two lists and
+      // removes events that are attended by people not invited to the meeting. 
+      if (Collections.disjoint(event.getAttendees(), request.getAttendees())) {
+        // Events with only optional attendees are added to optionalEvents. 
+        if (!Collections.disjoint(event.getAttendees(), request.getOptionalAttendees())) {
+          optionalEvents.add(event);
+        }
+        // Events that don't have any mandatory attendees are removed from sortedEvents.
+        eventsIterator.remove();
+      }
+    }
```

Comment:
There are two things happening in this block. One is adding optionalEvents and the other is modifying sortedEvents. Both list represent the same thing `list of events that contains at least 1 attendee from a list of attendees`. So you could make a helper method that takes all events and list of attendees, have it create a new list of events, and return it. In this main method, you call it twice. One with mandatory attendees and one with optional attendees. 

This removes the complicated mixture of two tasks and also the Java iterator usage.

For time efficiency, it is the same:
```
for(...) {
 // do x
 // do y
}
```
Is the same as 
```
for (...) {
  // do x
}
for (...) {
  // do y
}
```
(ignoring the negligible for loop overhead)

For space complexity, it is also the same cause in the end you just end up with two lists.

Reply:
I added the helper method to do this and my code's a lot cleaner now! Unfortunately I had to create a new list for mandatory attendee events because optionalEvents is dependent on it.